Doc/library/typing.rst

```diff
@@ -1266,213 +1323,327 @@ These can be used as types in annotations using ``[]``, each having a unique syn
-    semantics in several important ways. Using a *bound* type variable means
-    that the ``TypeVar`` will be solved using the most specific type possible::
+   Type variables can be bound to concrete types, abstract types (ABCs or
+   protocols), and even unions of types::
 
-       x = print_capitalized('a string')
-       reveal_type(x)  # revealed type is str
+      # Can be anything with an __abs__ method
+      def print_abs[T: SupportsAbs](arg: T) -> None:
+          print("Absolute value:", abs(arg))
 
-       class StringSubclass(str):
-           pass
+      U = TypeVar('U', bound=str|bytes)  # Can be any subtype of the union str|bytes
+      V = TypeVar('V', bound=SupportsAbs)  # Can be anything with an __abs__ method
 
-       y = print_capitalized(StringSubclass('another string'))
-       reveal_type(y)  # revealed type is StringSubclass
+   .. _typing-constrained-typevar:
 
-       z = print_capitalized(45)  # error: int is not a subtype of str
+   Using a *constrained* type variable, however, means that the ``TypeVar``
+   can only ever be solved as being exactly one of the constraints given::
 
-    Type variables can be bound to concrete types, abstract types (ABCs or
-    protocols), and even unions of types::
+      a = concatenate('one', 'two')
+      reveal_type(a)  # revealed type is str
 
-       U = TypeVar('U', bound=str|bytes)  # Can be any subtype of the union str|bytes
-       V = TypeVar('V', bound=SupportsAbs)  # Can be anything with an __abs__ method
+      b = concatenate(StringSubclass('one'), StringSubclass('two'))
+      reveal_type(b)  # revealed type is str, despite StringSubclass being passed in
 
-.. _typing-constrained-typevar:
+      c = concatenate('one', b'two')  # error: type variable 'A' can be either str or bytes in a function call, but not both
 
-    Using a *constrained* type variable, however, means that the ``TypeVar``
-    can only ever be solved as being exactly one of the constraints given::
+   At runtime, ``isinstance(x, T)`` will raise :exc:`TypeError`.
 
-       a = concatenate('one', 'two')
-       reveal_type(a)  # revealed type is str
+   .. attribute:: __name__
+
+      The name of the type variable.
+
+   .. attribute:: __covariant__
+
+      Whether the type var has been explicitly marked as covariant.
+
+   .. attribute:: __contravariant__
+
+      Whether the type var has been explicitly marked as contravariant.
+
+   .. attribute:: __infer_variance__
+
+      Whether the type variable's variance should be inferred by type checkers.
+
+      .. versionadded:: 3.12
+
+   .. attribute:: __bound__
+
+      The bound of the type variable, if any.
+
+      .. versionchanged:: 3.12
+
+         For type variables created through :ref:`type parameter syntax <type-params>`,
+         the bound is evaluated only when the attribute is accessed, not when
+         the type variable is created (see :ref:`lazy-evaluation`).
+
+   .. attribute:: __constraints__
```

Comment:
All of these except for `__infer_variance__` already exist on Python <=3.11. Maybe it would be better to add them in a separate PR, so that we can backport them?

Reply:
Same comment as above